flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/JsonUnquoteFunction.java

```diff
@@ -0,0 +1,128 @@
+        int i = 0;
+        while (i < inputStr.length()) {
+            if (inputStr.charAt(i) == '\\' && i + 1 < inputStr.length()) {
+                i++; // move to the next char
+                char ch = inputStr.charAt(i++);
+
+                switch (ch) {
+                    case '"':
+                        result.append(ch);
+                        break;
+                    case '\\':
+                        result.append(ch);
+                        break;
+                    case '/':
+                        result.append(ch);
+                        break;
+                    case 'b':
+                        result.append('\b');
+                        break;
+                    case 'f':
+                        result.append('\f');
+                        break;
+                    case 'n':
+                        result.append('\n');
+                        break;
+                    case 'r':
+                        result.append('\r');
+                        break;
+                    case 't':
+                        result.append('\t');
+                        break;
+                    case 'u':
+                        result.append(fromUnicodeLiteral(inputStr, i));
+                        i = i + 4;
+                        break;
+                    default:
+                        throw new RuntimeException("Illegal escape sequence: \\" + ch);
+                }
+            } else {
+                result.append(inputStr.charAt(i));
+                i++;
+            }
+        }
+        return result.toString();
+    }
+
+    private String unescapeValidJson(String inputStr) {
+        // check for a quoted json string val and unescape
+        if (inputStr.charAt(0) == '"' && inputStr.charAt(inputStr.length() - 1) == '"') {
+            // remove quotes, string len is atleast 2 here
+            return unescapeStr(inputStr.substring(1, inputStr.length() - 1));
+        } else {
+            // string representing Json - array, object or unquoted scalar val, return as-is
+            return inputStr;
+        }
+    }
+
+    private static String fromUnicodeLiteral(String input, int curPos) {
+
+        StringBuilder number = new StringBuilder();
+        // isValidJsonVal will already check for unicode literal validity
+        for (char ch : input.substring(curPos, curPos + 4).toCharArray()) {
+            number.append(Character.toLowerCase(ch));
+        }
+        int code = Integer.parseInt(number.toString(), 16);
+        return String.valueOf((char) code);
+    }
+
+    public @Nullable Object eval(Object input) {
+        if (input == null) {
+            return null;
+        }
+        BinaryStringData bs = (BinaryStringData) input;
+        String inputStr = bs.toString();
+        try {
+            if (isValidJsonVal(inputStr)) {
+                return new BinaryStringData(unescapeValidJson(inputStr));
+            }
+        } catch (Throwable t) {
```

Comment:
I don't think we should swallow any kind of exception or error.
Let's catch the errors that we expect (`IllegalArgumentException`, etc.) and handle those by returning the original input.
Unexpected exceptions should be forwarded because they might be related to other issues.

Reply:
sorry, commented on an outdated version...